Upgrade bindgen

This crate is currently using an old version of bindgen which requires LLVM 3.8 and no newer. The API and behavior has changed significantly since then, so it'll be a bit of work to upgrade.

I worked on this issue in the last few days, here is a status update on my progress so far. I would like to discuss potential approaches to deal with issues I encountered. You can find my work branch at https://github.com/ekse/rust-mbedtls/tree/bindgen-upgrade.

First, a quick note on why I want to upgrade. The current master branch does not compile on my computers, it fails with the error `Bindgen ERROR: unsupported type CXTypeKind(119)`. Upgrading bindgen fixes this issue. I get this build error on libclang 3.9, 4.0 and 6.0.

### "mbedtls_" prefix
To deal with the "mbedtls_" prefix, I tried the approach suggested by the bindgen maintainer to rename the imports with `use x as y` statements. To reduce the amount of boilerplate,  I experimented with creating a procedural macro "mbedtls_use" https://github.com/ekse/rust-mbedtls/tree/bindgen-upgrade/mbedtls-use. An example use of this macro looks like this: (https://github.com/ekse/rust-mbedtls/blob/bindgen-upgrade/mbedtls/src/cipher/raw/serde.rs#L24)

```rust
#[mbedtls_use]
use {
    mbedtls_aes_context, mbedtls_cipher_context_t, mbedtls_cipher_id_t, mbedtls_cipher_mode_t,
    mbedtls_des3_context, mbedtls_des_context, MBEDTLS_CIPHER_ID_3DES, MBEDTLS_CIPHER_ID_AES,
    MBEDTLS_CIPHER_ID_DES, MBEDTLS_MODE_CBC, MBEDTLS_MODE_CFB, MBEDTLS_MODE_CTR,
};
```

There are a couple downsides to this approach. Procedural macros were just stabilized in Rust 1.30, that would make it a hard requirement. Also, the use statements are not in scope when using macros like define!, so I used the full type names (see https://github.com/ekse/rust-mbedtls/blob/bindgen-upgrade/mbedtls/src/pk/ec.rs#L19).

Alternatively, we could remove the proc_macro and write the complete `use x as y` statements, at the cost of longer use statements.

### unsigned constants
The second issue I encountered is with constants. Bindgen defines constants such as `MBEDTLS_SSL_IS_CLIENT` as u32. This fails to compile for define! statements that map to c_int as it maps to i32. I assume the older version of bindgen defined those constants as i32.

```
error[E0308]: match arms have incompatible types                                                                                            
  --> mbedtls/src/wrapper_macros.rs:75:5                                                                                                    
   |                                                                                                                                        
75 |                   match self {                                                                                                         
   |  _________________^                                                                                                                    
76 | |                     $($n::$rust => ::mbedtls_sys::$c,)*                                                                              
   | |                                    ----------------- match arm with an incompatible type                                             
77 | |                 }                                                                                                                    
   | |_________________^ expected i32, found u32                                                                                            
   |                                                                                                                                        
  ::: mbedtls/src/ssl/config.rs:62:1                                                                                                        
   |                                                                                                                                        
62 | / define!(enum UseSessionTickets -> c_int {                                                                                            
63 | |     Enabled => MBEDTLS_SSL_SESSION_TICKETS_ENABLED,                                                                                  
64 | |     Disabled => MBEDTLS_SSL_SESSION_TICKETS_DISABLED,                                                                                
65 | | });                                                                                                                                  
   | |___- in this macro invocation  
```
Also, some getters and setters fail to compile because enums do not implement From conversions for i32 (because values in u32 could be outside the range of i32).

```
setter!(set_session_tickets(u: UseSessionTickets) = mbedtls_ssl_conf_session_tickets)

error[E0277]: the trait bound `i32: std::convert::From<ssl::config::UseSessionTickets>` is not satisfied                                    
   --> mbedtls/src/wrapper_macros.rs:192:50                                                                                                 
    |                                                                                                                                       
192 |             unsafe{::mbedtls_sys::$cfn(&mut self.inner,$n.into())}                                                                    
    |                                                           ^^^^ the trait `std::convert::From<ssl::config::UseSessionTickets>` is not implemented for `i32`

```
I temporarily worked around this by manually defining the constants as i32, and the crate successfully compiles at that point. To could be an approach to deal with this issue, I'm not sure what the best way to do that with bindgen would be.

---

Sorry for the long message, it seems upgrading the bindgen dependency will require a good amount of effort. I suggest we decide on a way to deal with the mbedtls_ prefix issue and move on from there.

Thanks,

Sébastien

--- a/.travis.yml
+++ b/.travis.yml
@@ -11,22 +11,26 @@ os: linux
 dist: xenial
 addons:
   apt:
+    sources:
+    - sourceline: "deb http://apt.llvm.org/xenial/ llvm-toolchain-xenial-11 main"
+      key_url: "https://apt.llvm.org/llvm-snapshot.gpg.key"
     packages:
-    - llvm-3.8-dev
-    - libclang-3.8-dev
-    - clang-3.8
+    - llvm-3.9-dev
+    - libclang-3.9-dev
+    - clang-3.9
+    - clang-11
     - cmake
 rust:
    - stable
    - beta
    - nightly
-   - nightly-2019-07-01
+   - nightly-2021-03-25
 env:
   global:
     - RUST_BACKTRACE=1
     # Pinned to this particular nightly version because of core_io. This can be
     # re-pinned whenever core_io is updated to the latest nightly.
-    - CORE_IO_NIGHTLY=nightly-2019-07-01
-    - LLVM_CONFIG_PATH=llvm-config-3.8
+    - CORE_IO_NIGHTLY=nightly-2021-03-25
+    - LLVM_CONFIG_PATH=llvm-config-3.9
 script:
   - ./ct.sh

--- a/ct.sh
+++ b/ct.sh
@@ -9,6 +9,9 @@ if [ -z $TRAVIS_RUST_VERSION ]; then
     exit 1
 fi
 
+export CFLAGS_x86_64_fortanix_unknown_sgx="-isystem/usr/include/x86_64-linux-gnu -mlvi-hardening -mllvm -x86-experimental-lvi-inline-asm-hardening"
+export CC_x86_64_fortanix_unknown_sgx=clang-11
+
 if [ $TRAVIS_RUST_VERSION = "stable" ] || [ $TRAVIS_RUST_VERSION = "beta" ] || [ $TRAVIS_RUST_VERSION = "nightly" ]; then
     rustup default $TRAVIS_RUST_VERSION
     # make sure that explicitly providing the default target works

--- a/mbedtls-sys/Cargo.toml
+++ b/mbedtls-sys/Cargo.toml
@@ -27,9 +27,12 @@ libc = { version = "0.2.0", optional = true }
 libc = { version = "0.2.0" }
 
 [build-dependencies]
-bindgen = "0.19.0"
+bindgen = "0.58.0"
 cmake = "0.1.17"
+cc = "1.0.45"
 lazy_static = "1.4"
+syn = { version = "1.0.64", features = ["full", "visit"] }
+quote = "1.0.9"
 
 [features]
 # If you use mbedtls-sys in a no_std configuration, you need to provide your

--- a/mbedtls-sys/build/bindgen.rs
+++ b/mbedtls-sys/build/bindgen.rs
@@ -8,78 +8,132 @@
 
 use bindgen;
 
-use std::fs::File;
-use std::io::{stderr, Write};
+use std::fmt::Write as _;
+use std::fs::{self, File};
+use std::io::Write;
 
 use crate::headers;
 
 #[derive(Debug)]
-struct StderrLogger;
+struct MbedtlsParseCallbacks;
 
-impl bindgen::Logger for StderrLogger {
-    fn error(&self, msg: &str) {
-        let _ = writeln!(stderr(), "Bindgen ERROR: {}", msg);
+impl bindgen::callbacks::ParseCallbacks for MbedtlsParseCallbacks {
+    fn item_name(&self, original_item_name: &str) -> Option<String> {
+        Some(original_item_name.trim_start_matches("mbedtls_").trim_start_matches("MBEDTLS_").to_owned())
     }
-    fn warn(&self, msg: &str) {
-        let _ = writeln!(stderr(), "Bindgen WARNING: {}", msg);
+
+    fn enum_variant_name(
+        &self,
+        _enum_name: Option<&str>,
+        original_variant_name: &str,
+        _variant_value: bindgen::callbacks::EnumVariantValue
+    ) -> Option<String> {
+        self.item_name(original_variant_name)
     }
+
+    fn int_macro(&self, _name: &str, value: i64) -> Option<bindgen::callbacks::IntKind> {
+        if value < (i32::MIN as i64) || value > (i32::MAX as i64) {
+            Some(bindgen::callbacks::IntKind::LongLong)
+        } else {
+            Some(bindgen::callbacks::IntKind::Int)
+        }
+    }
+
+    fn blocklisted_type_implements_trait(&self, _name: &str, derive_trait: bindgen::callbacks::DeriveTrait) -> Option<bindgen::callbacks::ImplementsTrait> {
+        if derive_trait == bindgen::callbacks::DeriveTrait::Default {
+            Some(bindgen::callbacks::ImplementsTrait::Manually)
+        } else {
+            Some(bindgen::callbacks::ImplementsTrait::Yes)
+        }
+    }
+}
+
+/// Add bindgen 0.19-style union accessor methods. These are deprecated
+/// and can be deleted with the next major version bump.
+fn generate_deprecated_union_accessors(bindings: &str) -> String {
+    #[derive(Default)]
+    struct UnionImplBuilder {
+        impls: String
+    }
+
+    impl<'ast> syn::visit::Visit<'ast> for UnionImplBuilder {
+        fn visit_item_union(&mut self, i: &'ast syn::ItemUnion) {
+            let union_name = &i.ident;
+            let field_name = i.fields.named.iter().map(|field| field.ident.as_ref().unwrap());
+            let field_type = i.fields.named.iter().map(|field| &field.ty);
+            write!(self.impls, "{}", quote::quote! {
+                impl #union_name {
+                    #(
+                        #[deprecated]
+                        pub unsafe fn #field_name(&mut self) -> *mut #field_type {
+                            &mut self.#field_name
+                        }
+                    )*
+                }
+            }).unwrap();
+        }
+    }
+
+    let mut impl_builder = UnionImplBuilder::default();
+    syn::visit::visit_file(&mut impl_builder, &syn::parse_file(&bindings).unwrap());
+
+    impl_builder.impls
 }
 
 impl super::BuildConfig {
     pub fn bindgen(&self) {
-        let header = self.out_dir.join("bindgen-input.h");
-        File::create(&header)
-            .and_then(|mut f| {
-                Ok(for h in headers::enabled_ordered() {
-                    writeln!(f, "#include <mbedtls/{}>", h)?;
-                })
-            }).expect("bindgen-input.h I/O error");
+        let mut input = String::new();
+        for h in headers::enabled_ordered() {
+            let _ = writeln!(input, "#include <mbedtls/{}>", h);
+        }
 
-        let include = self.mbedtls_src.join("include");
+        let mut cc = cc::Build::new();
+        cc.include(&self.mbedtls_include)
+        .flag(&format!(
+            "-DMBEDTLS_CONFIG_FILE=\"{}\"",
+            self.config_h.to_str().expect("config.h UTF-8 error")
+        ));
 
-        let logger = StderrLogger;
-        let mut bindgen = bindgen::Builder::new(header.into_os_string().into_string().unwrap());
-        let bindings = bindgen
-            .log(&logger)
-            .clang_arg("-Dmbedtls_t_udbl=mbedtls_t_udbl;") // bindgen can't handle unused uint128
-            .clang_arg(format!(
-                "-DMBEDTLS_CONFIG_FILE=\"{}\"",
-                self.config_h.to_str().expect("config.h UTF-8 error")
-            )).clang_arg(format!(
-                "-I{}",
-                include.to_str().expect("include/ UTF-8 error")
-            )).match_pat(include.to_str().expect("include/ UTF-8 error"))
-            .match_pat(self.config_h.to_str().expect("config.h UTF-8 error"))
-            .use_core(true)
+        for cflag in &self.cflags {
+            cc.flag(cflag);
+        }
+
+        let bindings = bindgen::builder()
+            .clang_args(cc.get_compiler().args().iter().map(|arg| arg.to_str().unwrap()))
+            .header_contents("bindgen-input.h", &input)
+            .allowlist_function("^(?i)mbedtls_.*")
+            .allowlist_type("^(?i)mbedtls_.*")
+            .allowlist_var("^(?i)mbedtls_.*")
+            .allowlist_recursively(false)
+            .blocklist_type("^mbedtls_time_t$")
+            .use_core()
+            .ctypes_prefix("::types::raw_types")
+            .parse_callbacks(Box::new(MbedtlsParseCallbacks))
+            .default_enum_style(bindgen::EnumVariation::Consts)
+            .generate_comments(false)
+            .derive_copy(true)
             .derive_debug(false) // buggy :(
-            .ctypes_prefix(vec!["types".to_owned(), "raw_types".to_owned()])
-            .remove_prefix("mbedtls_")
-            .rust_enums(false)
-            .convert_macros(true)
-            .macro_int_types(
-                vec![
-                    "sint",
-                    "sint",
-                    "sint",
-                    "slonglong",
-                    "sint",
-                    "sint",
-                    "sint",
-                    "slonglong",
-                ].into_iter(),
-            ).generate()
-            .expect("bindgen error");
+            .derive_default(true)
+            .prepend_enum_name(false)
+            .translate_enum_integer_types(true)
+            .rustfmt_bindings(false)
+            .raw_line("#![allow(dead_code, non_snake_case, non_camel_case_types, non_upper_case_globals, invalid_value)]")
+            .generate()
+            .expect("bindgen error")
+            .to_string();
+
+        let union_impls = generate_deprecated_union_accessors(&bindings);
 
         let bindings_rs = self.out_dir.join("bindings.rs");
         File::create(&bindings_rs)
             .and_then(|mut f| {
-                bindings.write(Box::new(&mut f))?;
-                f.write_all(b"use crate::types::*;\n") // for FILE, time_t, etc.
+                f.write_all(bindings.as_bytes())?;
+                f.write_all(union_impls.as_bytes())?;
+                f.write_all(b"use crate::types::*;\n")?; // for FILE, time_t, etc.
+                Ok(())
             }).expect("bindings.rs I/O error");
 
         let mod_bindings = self.out_dir.join("mod-bindings.rs");
-        File::create(&mod_bindings)
-            .and_then(|mut f| f.write_all(b"mod bindings;\n"))
-            .expect("mod-bindings.rs I/O error");
+        fs::write(mod_bindings, b"mod bindings;\n").expect("mod-bindings.rs I/O error");
     }
 }

--- a/mbedtls-sys/build/build.rs
+++ b/mbedtls-sys/build/build.rs
@@ -28,7 +28,9 @@ use features::FEATURES;
 struct BuildConfig {
     out_dir: PathBuf,
     mbedtls_src: PathBuf,
+    mbedtls_include: PathBuf,
     config_h: PathBuf,
+    cflags: Vec<String>,
 }
 
 impl BuildConfig {
@@ -87,17 +89,32 @@ impl BuildConfig {
             );
         }
     }
+
+    fn new() -> Self {
+        let out_dir = PathBuf::from(env::var_os("OUT_DIR").expect("OUT_DIR environment not set?"));
+        let config_h = out_dir.join("config.h");
+        let mbedtls_src = PathBuf::from(env::var("RUST_MBEDTLS_SYS_SOURCE").unwrap_or("vendor".to_owned()));
+        let mbedtls_include = mbedtls_src.join("include");
+
+        let mut cflags = vec![];
+        if FEATURES.have_platform_component("c_compiler", "freestanding") {
+            cflags.push("-fno-builtin".into());
+            cflags.push("-D_FORTIFY_SOURCE=0".into());
+            cflags.push("-fno-stack-protector".into());
+        }
+
+        BuildConfig {
+            config_h,
+            out_dir,
+            mbedtls_src,
+            mbedtls_include,
+            cflags,
+        }
+    }
 }
 
 fn main() {
-    let out_dir = PathBuf::from(env::var_os("OUT_DIR").expect("OUT_DIR environment not set?"));
-    let src = PathBuf::from(env::var("RUST_MBEDTLS_SYS_SOURCE").unwrap_or("vendor".to_owned()));
-    let cfg = BuildConfig {
-        config_h: out_dir.join("config.h"),
-        out_dir: out_dir,
-        mbedtls_src: src,
-    };
-
+    let cfg = BuildConfig::new();
     cfg.create_config_h();
     cfg.print_rerun_files();
     cfg.cmake();

--- a/mbedtls-sys/build/cmake.rs
+++ b/mbedtls-sys/build/cmake.rs
@@ -8,8 +8,6 @@
 
 use cmake;
 
-use crate::features::FEATURES;
-
 impl super::BuildConfig {
     pub fn cmake(&self) {
         let mut cmk = cmake::Config::new(&self.mbedtls_src);
@@ -20,12 +18,16 @@ impl super::BuildConfig {
         .define("ENABLE_PROGRAMS", "OFF")
         .define("ENABLE_TESTING", "OFF")
         .build_target("lib");
-        if FEATURES.have_platform_component("c_compiler", "freestanding") {
-            cmk.cflag("-fno-builtin")
-                .cflag("-D_FORTIFY_SOURCE=0")
-                .cflag("-fno-stack-protector");
+        for cflag in &self.cflags {
+            cmk.cflag(cflag);
         }
+        let cc = cc::Build::new().get_compiler();
+        if cc.is_like_clang() && cc.args().iter().any(|arg| arg == "-mllvm") {
+            cmk.define("CMAKE_C_COMPILER_FORCED", "TRUE");
+        }
+
         let mut dst = cmk.build();
+
         dst.push("build");
         dst.push("library");
         println!(
@@ -33,8 +35,7 @@ impl super::BuildConfig {
             dst.to_str().expect("link-search UTF-8 error")
         );
 
-        let mut dst = cmk.build();
-        dst.push("build");
+        assert!(dst.pop());
         dst.push("crypto");
         dst.push("library");
         println!(
@@ -46,7 +47,7 @@ impl super::BuildConfig {
         println!("cargo:rustc-link-lib=mbedx509");
         println!("cargo:rustc-link-lib=mbedcrypto");
 
-        println!("cargo:include={}/{}", ::std::env::current_dir().unwrap().display(), self.mbedtls_src.join("include").display());
+        println!("cargo:include={}", ::std::env::current_dir().unwrap().join(&self.mbedtls_include).to_str().expect("include/ UTF-8 error"));
         println!("cargo:config_h={}", self.config_h.to_str().expect("config.h UTF-8 error"));
     }
 }

--- a/mbedtls-sys/build/config.rs
+++ b/mbedtls-sys/build/config.rs
@@ -413,7 +413,6 @@ pub fn default_defines() -> HashMap<&'static str, Macro> {
 pub const FEATURE_DEFINES: &'static [(&'static str, CDefine)] = &[
     ("time",                  ("MBEDTLS_HAVE_TIME",                         Defined)),
     ("time",                  ("MBEDTLS_HAVE_TIME_DATE",                    Defined)),
-    ("time",                  ("MBEDTLS_TIMING_C",                          Defined)),
     ("havege",                ("MBEDTLS_HAVEGE_C",                          Defined)),
     ("threading",             ("MBEDTLS_THREADING_C",                       Defined)),
     ("pkcs11",                ("MBEDTLS_PKCS11_C",                          Defined)),
@@ -437,6 +436,7 @@ pub const FEATURE_DEFINES: &'static [(&'static str, CDefine)] = &[
 
 #[cfg_attr(rustfmt, rustfmt_skip)]
 pub const PLATFORM_DEFINES: &'static [(&'static str, &'static str, CDefine)] = &[
+    ("time",      "libc",     ("MBEDTLS_TIMING_C",                          Defined)),
     ("time",      "custom",   ("MBEDTLS_PLATFORM_TIME_MACRO",               DefinedAs("mbedtls_time"))),
     ("time",      "custom",   ("MBEDTLS_PLATFORM_TIME_TYPE_MACRO",          DefinedAs("long long"))),
     ("time",      "custom",   ("MBEDTLS_PLATFORM_GMTIME_R_ALT",             Defined)),

